### PR TITLE
[package] ignore dunder import errors

### PR DESCRIPTION
**Pull request by lobstergrindset**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61148**
* #61147
* #61146

Changes `__import__` processing to silently skip cases where the `__import__` statement cannot be parsed

Differential Revision: [D29559680](https://our.internmc.facebook.com/intern/diff/D29559680)

### CODE REVIEW

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559680).

File: torch/package/package_exporter.py [hunk 1]
```diff
@@ -970,6 +974,15 @@ def relied_on_by(self, module_name: str) -> List[str]:
         else:
             return []
 
+    def failed_dunder_import_list(self) -> List[str]:
```
Comment (suo) [inline torch/package/package_exporter.py:977]: I'm not sure what discoverability for this API will be. Ideally we'd like the API to be "obvious"—someone will look in the API reference pretty and quickly find what they're looking for. The methods introduced in the PR below are good examples of obvious APIs—they correspond to actions that are common in the rest the API and it's pretty clear what they do even from the name.

But `failed_dunder_import_list` is pretty inscrutable—I'm not sure how a user would figure out what this method does, or how they would know to call it, or what they would do with results. Maybe we should just eat the `__import__` error and call it a day?

**Comment by Lilyjjo** [comment]
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29559680).

File: torch/package/find_file_dependencies.py [hunk 2]
```diff
@@ -57,47 +57,46 @@ def _grab_node_str(self, node):
     def visit_Call(self, node):
         # __import__ calls aren't routed to the visit_Import/From nodes
         if hasattr(node.func, "id") and node.func.id == "__import__":
-            if type(node.args[0]) not in [ast.Constant, ast.Str]:
-                # We don't want to parse dynamic uses of __import__
+            try:
```
Comment (suo) [inline torch/package/find_file_dependencies.py:60]: can we narrow the check? we really only want to catch when the imported module isn't constant, any other exceptions should get raised properly. It seems like the existing check only checks the first argument for constantness—maybe the narrowest check would be to check all arguments for constantness and skip if any of them aren't.

**Review by suo — CHANGES_REQUESTED** [review]

File: torch/package/find_file_dependencies.py (hunk 2 above)
Comment (Lilyjjo) [reply torch/package/find_file_dependencies.py:60]: That might be too strict of a check -- The global/local variables aren't constants normally ex: `module = __import__(module_name, globals(), locals(), [''])`

**Review by suo — APPROVED** [review]

File: torch/package/find_file_dependencies.py (hunk 2 above)
Comment (suo) [reply torch/package/find_file_dependencies.py:60]: alright, it's probably not so bad anyway since the try-catch only covers `__import__` calls

**Pull request closed without merging**